content/post/fewer-parameters-means-fewer-bugs.md

```diff
@@ -0,0 +1,244 @@
+
+```rb
+class Book
+  attr_reader :title, :year_published, :author
+  
+  def initialize(title, year_published, author)
+    @title = title
+    @year_published = year_published
+    @author = author
+  end
+ 
+  def summary
+    "#{title} - #{author.full_name} (#{year_published})"
+  end
+  
+  def cover
+    "#{title}\n\n\n#{author.full_name}"
+  end
+end
+  
+bater = Author.new("Bater", "Chen")
+book = Book.new("Ruby is the best", "2020", bater)
+book.summary
+=> "Ruby is the best - Chen, Bater (2020)"
+book.cover
+=> "Ruby is the best\n\n\nChen, Bater"
+```
+Both `summary` and `cover` methods' output remain the same but now we can call it freely without any parameter once the book object is prepared. In Ruby, we can skip parentheses when there is no parameter.
+
+This example doesn't encourage you to move all the parameters into the object attributes, it depends on the domain know-how. We can do it because it makes sense for a book object to having an author as an attribute in this case. It may get worse if we try to put the wrong data as the object attribute.
+
+## Open-closed principle
+When we reducing our parameters from two to zero, our design is now open for extension but closed for modification. This is inline with the [open-closed principle][3].
+
+Let's say we have a new requirement again, this time, some authors now have a middle name. In this new version, it would be very easy to adapt our code because we don't need to touch the book class anymore. The new author class would refactored like this:
+
+```rb
+class Author
+  attr_reader :first_name, :last_name, :middle_name
+  
+  def initialize(first_name, last_name, middle_name = nil)
+    @first_name = first_name
+    @last_name = last_name
+    @middle_name = middle_name
+  end
+  
+  def full_name
+    if middle_name.nil?
+      "#{last_name}, #{first_name}"
+    else
+      "#{last_name}, #{middle_name}, #{first_name}"
+    end
+  end
+end
+  
+bater = Author.new("Bater", "Chen")
+Book.new("Ruby is the best", "2020", bater).summary
+=> "Ruby is the best - Chen, Bater (2020)"
+  
+robert = Author.new("Robert", "Kiyosaki", "T")
+Book.new("Rich Dad, Poor Dad", "1997", robert).summary
+=> "Rich Dad, Poor Dad - Kiyosaki, T, Robert (1997)"
+```
+In this design, the existing behavior doesn't change. The default value of the middle name is `nil` (also known as Null in other languages) because in most cases we can skip it and we won't break anything when the existing code tries to create a new author with only two parameters. The author class is open for extension and the book class is closed for modification in this case. Without refactoring, we need to touch upon both the `summary` and `cover` methods, causing the parameters to increase from two to three. Let's compare the code before and after, again:
```

Comment:
```suggestion
In this design, the existing behavior doesn't change. The default value of the middle name is `nil` (also known as `null` in other languages) because in most cases we can skip it and we won't break anything when the existing code tries to create a new author with only two parameters. The `Author` class is _open for extension_ and the `Book` class is _closed for modification_ in this case. Without refactoring, we need to touch upon both the `summary` and `cover` methods, causing the parameters to increase from two to three. Let's compare the code before and after, again:
```